hphp/runtime/vm/jit/vasm-reuseimm.cpp

```diff
@@ -0,0 +1,161 @@
+   +----------------------------------------------------------------------+
+*/
+
+#include "hphp/runtime/vm/jit/vasm.h"
+
+#include "hphp/runtime/vm/jit/vasm-gen.h"
+#include "hphp/runtime/vm/jit/vasm-instr.h"
+#include "hphp/runtime/vm/jit/vasm-print.h"
+#include "hphp/runtime/vm/jit/vasm-reg.h"
+#include "hphp/runtime/vm/jit/vasm-unit.h"
+#include "hphp/runtime/vm/jit/vasm-util.h"
+
+#include "hphp/util/arch.h"
+#include "hphp/util/assertions.h"
+
+#include <folly/Optional.h>
+
+#include <cstdlib>
+
+TRACE_SET_MOD(vasm);
+
+namespace HPHP { namespace jit {
+
+namespace {
+
+// track ldimmq values
+struct ImmState {
+  ImmState() : val{0}, base{} {}
+  ImmState(Immed64 a, Vreg b) : val{a}, base{b} {}
+
+  void reset() { base = Vreg{}; }
+
+  Immed64 val;
+  Vreg base;
+};
+
+struct Env {
+  Vunit& unit;
+  std::vector<ImmState> immStateVec;
+};
+
+bool isMultiword(int64_t imm) {
+  switch (arch()) {
+  case Arch::X64:
+  case Arch::PPC64:
+    break;
+  case Arch::ARM:
+    uint64_t val = std::abs(imm);
+    if (val > (1 << 16)) return true;
+    break;
+  }
+  return false;
+}
+
+// candidate around +/- uimm12
+folly::Optional<int> reuseCandidate(Env& env, int64_t p, Vreg& reg) {
+  for (auto const& elem : env.immStateVec) {
+    if (!elem.base.isValid()) continue;
+    int64_t q = elem.val.q();
+    if ((p >= q) && (p < (q + 4095))) {
+      reg = elem.base;
+      return folly::Optional<int>(safe_cast<int>(p - q));
+    }
+    if ((p < q) && (q < (p + 4095))) {
+      reg = elem.base;
+      return folly::Optional<int>(safe_cast<int>(p - q));
+    }
+  }
+  return folly::none;
+}
+
+template <typename Inst>
+void reuseImmq(Env& env, const Inst& inst, Vlabel b, size_t i) {
+  // leaky bucket
+  env.immStateVec[i % RuntimeOption::EvalJitLdimmqSpan] = ImmState{};
+}
+
+template<typename ReuseImm>
+void reuseimm_impl(Vunit& unit, Vlabel b, size_t i, ReuseImm reuse) {
+  vmodify(unit, b, i, [&] (Vout& v) { reuse(v); return 1; });
+}
+
+void reuseImmq(Env& env, const ldimmq& ld, Vlabel b, size_t i) {
+  if (isMultiword(ld.s.q())) {
+    Vreg base;
+    auto const off = reuseCandidate(env, ld.s.q(), base);
+
+    if (off.hasValue()) {
+      reuseimm_impl(env.unit, b, i, [&] (Vout& v) {
+        v << addqi{off.value(), base, ld.d, v.makeReg()};
+      });
+      return;
+    }
+  }
+  env.immStateVec[i % RuntimeOption::EvalJitLdimmqSpan] = ImmState{ld.s, ld.d};
+}
+
+void reuseImmq(Env& env, Vlabel b, size_t i) {
+  assertx(i <= env.unit.blocks[b].code.size());
+  auto& inst = env.unit.blocks[b].code[i];
+
+  if (isBlockEnd(inst) || isCall(inst)) {
+    for (auto &elem : env.immStateVec) elem.reset();
+    return;
+  } 
+
+  switch (inst.op) {
+#define O(name, ...)                      \
+    case Vinstr::name:                    \
+      reuseImmq(env, inst.name##_, b, i); \
+      break;
+
+    VASM_OPCODES
+#undef O
+  }
+}
+
+}
+
+// Opportunistically reuse immediate q values
+void reuseImmq(Vunit& unit) {
```

Comment:
@cmuellner - Hi Christoph, The intent of this pass was to reduce code size on ARMv8.  
Since we have to recreate some  immediates with multi-instruction sequences (movz/movk), it's checking a small leaky window to see if a close value is already in a register.  The window
size is optionable in case this increases register pressure.  Values which can be done in
one instruction are not considered.